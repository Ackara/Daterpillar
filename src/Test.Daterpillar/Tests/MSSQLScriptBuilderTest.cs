using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Acklann.Daterpillar.TextTransformation;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Tests.Daterpillar.Constants;
using Tests.Daterpillar.Helpers;

namespace Test.Daterpillar.Tests
{
    [TestClass]
    [DeploymentItem(SampleData.Folder)]
    [DeploymentItem(KnownFile.DbConfig)]
    [UseApprovalSubdirectory(nameof(ApprovalTests))]
    [UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
    public class MSSQLScriptBuilderTest : DbTemplateBuilderTestBase
    { 
        [ClassCleanup]
        public static void Cleanup()
        {
            ApprovalTests.Maintenance.ApprovalMaintenance.CleanUpAbandonedFiles();
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_builds_a_new_schema_when_settings_are_enabled()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                var settings = new ScriptBuilderSettings()
                {
                    AppendScripts = true,
                    AppendComments = true,
                    CreateDatabase = true,
                    TruncateDatabaseIfItExist = true
                };

                RunSchemaTest<MSSQLScriptBuilder>(settings, connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_builds_a_new_schema_when_settings_are_disabled()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                var settings = new ScriptBuilderSettings()
                {
                    AppendScripts = false,
                    AppendComments = false,
                    CreateDatabase = false,
                    TruncateDatabaseIfItExist = false
                };

                RunSchemaTestWithDisabledSettings<MSSQLScriptBuilder>(settings, connection);
            }
        }

        [TestMethod]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_adds_a_new_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunColumnTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_adds_a_new_index_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunIndexTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Create_should_generate_a_tsql_script_that_adds_a_new_foreign_key_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunForeignKeyTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_that_removes_a_schema_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunSchemaDropTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_that_removes_a_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunColumnDropTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_that_removes_a_table_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunTableDropTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_that_removes_a_index_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunIndexDropTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Drop_should_generate_a_tsql_script_that_removes_a_foreign_key_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunForeignKeyDropTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Alter_should_generate_a_tsql_modify_script_for_a_table_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunAlterTableTest<MSSQLScriptBuilder>(connection);
            }
        }

        [TestMethod]
        [Owner(Dev.Ackara)]
        [TestCategory(Trait.Integration)]
        public void Alter_should_generate_a_tsql_modify_script_for_a_column_when_invoked()
        {
            using (var connection = DatabaseHelper.CreateMSSQLConnection())
            {
                RunAlterColumnTest<MSSQLScriptBuilder>(connection);
            }
        }
    }
}