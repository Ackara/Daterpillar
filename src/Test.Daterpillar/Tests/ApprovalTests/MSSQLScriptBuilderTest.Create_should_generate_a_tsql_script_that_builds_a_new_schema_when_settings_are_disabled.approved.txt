IF OBJECT_ID('genre') IS NULL CREATE TABLE [genre]
(
	[Id] INT NOT NULL PRIMARY KEY IDENTITY(1, 1),
	[Name] VARCHAR(64) NOT NULL DEFAULT 'Other'
);


IF OBJECT_ID('artist') IS NULL CREATE TABLE [artist]
(
	[Id] INT NOT NULL PRIMARY KEY IDENTITY(1, 1),
	[Name] VARCHAR(64) NOT NULL,
	[Bio] TEXT
);

IF NOT EXISTS(SELECT * FROM [sys].[indexes] WHERE [object_id]=OBJECT_ID('artist') AND [name]='artist_idx1') CREATE UNIQUE INDEX [artist_idx1] ON [artist] ([Name] DESC);

IF OBJECT_ID('album') IS NULL CREATE TABLE [album]
(
	[Artist_Id] INT NOT NULL,
	[Name] VARCHAR(64) NOT NULL,
	[Release_Date] DATE,
	CONSTRAINT [artist_id_TO_artist_id_fkey2] FOREIGN KEY ([Artist_Id]) REFERENCES [artist]([Id]) ON UPDATE CASCADE ON DELETE CASCADE
);

ALTER TABLE [album] ADD PRIMARY KEY ([Artist_Id] ASC, [Name] ASC);

IF OBJECT_ID('song') IS NULL CREATE TABLE [song]
(
	[Id] INT NOT NULL PRIMARY KEY IDENTITY(1, 1),
	[Album_Id] INT NOT NULL,
	[Artist_Id] INT NOT NULL,
	[Genre_Id] INT NOT NULL,
	[Name] VARCHAR(256) NOT NULL,
	[Length] DECIMAL(4, 2) DEFAULT '0',
	[Price] DECIMAL(12, 2) NOT NULL,
	[On_Device] BIT NOT NULL,
	CONSTRAINT [genre_id_TO_genre_id_fkey4] FOREIGN KEY ([Genre_Id]) REFERENCES [genre]([Id]) ON UPDATE CASCADE ON DELETE CASCADE,
	CONSTRAINT [artist_id_TO_artist_id_fkey5] FOREIGN KEY ([Artist_Id]) REFERENCES [artist]([Id]) ON UPDATE CASCADE ON DELETE CASCADE
);

IF NOT EXISTS(SELECT * FROM [sys].[indexes] WHERE [object_id]=OBJECT_ID('song') AND [name]='song_idx') CREATE INDEX [song_idx] ON [song] ([Name] ASC);

GO
